### ISSUE TITLE
Embedding-E

### ISSUE THREAD

**Issue by Mahdi-122** [Mahdi-122/T-122#4]
from datasets import load_dataset
from scipy.sparse import lil_matrix
import numpy as np
from sklearn.decomposition import TruncatedSVD
from sklearn.manifold import TSNE
import plotly.express as px
import pandas as pd
import re
from collections import Counter

LawDataset = load_dataset("QomSSLab/law-text-dataset-fa")
LawDataTxt = LawDataset["train"]["text"]

def Normalize():
    global NomText
    NomText = []
    for txt in LawDataTxt :
        txt = txt.replace("ي","ی")
        txt = txt.replace("ك","ک")
        txt = txt.replace("ة","ه")
        txt = txt.replace("ؤ","و")
        txt = txt.replace("ئ","ی")
        txt = txt.replace("أ","ا")
        txt = txt.replace("إ","ا")
        txt = txt.replace("آ","ا")
        txt = txt.replace("-","")
        txt = txt.replace("،", "")
        txt = txt.replace("\n"," ")
        txt = txt.replace("\u200c"," ")
        txt = txt.replace("\u200f", "")
        txt = txt.replace("\u200e", "")
        txt = txt.replace("–", "")
        txt = txt.replace("\u200b", "")
        txt = txt.replace("\u200d", "")
        txt = txt.replace("►", "")
        txt = txt.replace("═", "")
        txt = txt.replace("≤", "")
        txt = txt.replace(".", "")
        txt = txt.replace("»", "")
        txt = txt.replace("«", "")
        txt = txt.replace("**","")
        txt = txt.replace("\"","")
        txt = txt.replace("۰","")
        txt = txt.replace("۱","")
        txt = txt.replace("۲","")
        txt = txt.replace("۳","")
        txt = txt.replace("۴","")
        txt = txt.replace("۵","")
        txt = txt.replace("۶","")
        txt = txt.replace("۷","")
        txt = txt.replace("۸","")
        txt = txt.replace("۹","")
        txt = txt.replace("#","")
        txt = txt.replace("&","")
        txt = txt.replace("…","")
        txt = txt.replace(")","")
        txt = txt.replace("(","")
        txt = txt.replace("!","")
        txt = txt.replace(">","")
        txt = txt.replace("<","")
        txt = re.sub(r'http[s]?://\S+|www\.\S+', '', txt)
        txt = txt.strip()

        NomText.append(txt)
    return NomText

Normalize()

def Tokenizer():
    Tokens = []
    for t in NomText:
        Tokens.append(t.split())
    return Tokens

Tokenized = Tokenizer()

all_words = [w for s in Tokenized for w in s]
unique_w = sorted(list(set(all_words)))
UW_size = len(unique_w)

Word_idx = {word: i for i, word in enumerate(unique_w)}
idx_Word = {i: word for word, i in Word_idx.items()}

print("Vocabulary size:", UW_size)

window_size = int(input("Window size: "))

Co_Matrix = lil_matrix((UW_size, UW_size), dtype=np.int32)

for s in Tokenized:
    inc = [Word_idx[w] for w in s if w in Word_idx]
    for i, center in enumerate(inc):
        start = max(0, i - window_size)
        end = min(len(inc), i + window_size + 1)
        for j in range(start, end):
            if i != j:
                Co_Matrix[center, inc[j]] += 1

print("Non-zero entries:", Co_Matrix.count_nonzero())

Dense_CM = Co_Matrix.toarray().astype(np.float32)
MeanCentered = Dense_CM - Dense_CM.mean(axis=0)

svd = TruncatedSVD(n_components=100)
Word_Embeddings = svd.fit_transform(MeanCentered)

print("Embedding shape:", Word_Embeddings.shape)

def word_similarity(w1, w2):
    if w1 not in Word_idx:
        return f"Unknown word: {w1}"
    if w2 not in Word_idx:
        return f"Unknown word: {w2}"

    i1 = Word_idx[w1]
    i2 = Word_idx[w2]

    v1 = Word_Embeddings[i1]
    v2 = Word_Embeddings[i2]

    num = np.dot(v1, v2)
    den = np.linalg.norm(v1) * np.linalg.norm(v2)

    if den == 0:
        return "Invalid vector."

    return num / den

wordC1 = input("Choose your what word you want to compare : ")
wordC2 = input("Choose the second your what word you want to compare : ")

print("Similarity:", word_similarity(wordC1, wordC2))

freq = Counter(all_words)
top_words = [w for w, c in freq.most_common(1000)]
top_indices = [Word_idx[w] for w in top_words]

top_emb = Word_Embeddings[top_indices]

tsne = TSNE(n_components=2, perplexity=30, learning_rate=200)
emb_2d = tsne.fit_transform(top_emb)

df = pd.DataFrame({
    "x": emb_2d[:, 0],
    "y": emb_2d[:, 1],
    "word": top_words
})

fig = px.scatter(df, x="x", y="y", text="word", width=1500, height=900)
fig.update_traces(textposition='top center')
fig.show()
